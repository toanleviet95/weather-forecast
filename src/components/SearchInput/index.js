import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import DOMPurify from 'dompurify';
import Spinner from '@/components/Spinner';
import useOutside from '@/shared/useOutside';
import useDebounce from '@/shared/useDebounce';
import useAsync from '@/shared/useAsync';
import locationService from '@/services/location.service';
import { DEBOUNCE_TIME } from '@/shared/constants';
import { Wrapper } from './index.styled';

const SearchInput = ({ onSelect }) => {
  const [data, setData] = useState([]);
  const [active, setActive] = useState(false);
  const [value, setValue] = useState('');
  const debouncedSearchTerm = useDebounce(value, DEBOUNCE_TIME);
  const wrapperRef = useRef(null);
  const { execute, value: locationValue, status } = useAsync(locationService, '', false);
  
  const handleFocus = () => {
    setActive(true);
  };

  const handleChange = (e) => {
    setValue(e.target.value);
    setData([]);
  };

  const handleReset = () => {
    setActive(false);
    setValue('');
    setData([]);
  };

  const handleSelect = ({title, woeid}) => {
    setValue(title);
    setActive(false);
    setData([]);
    onSelect(woeid);
  };

  useOutside(wrapperRef, () => {
    setActive(false);
  });

  useEffect(() => {
    if (debouncedSearchTerm && active) {
      setActive(true);
      // Fetch location data
      execute(debouncedSearchTerm);
    } else {
      setData([]);
    }
  }, [debouncedSearchTerm]);

  useEffect(() => {
    if (status === 'success' && debouncedSearchTerm) {
      const { data: hits } = locationValue;
      const regex = RegExp(debouncedSearchTerm, 'gi');
      const replacement = `<b><mark>${debouncedSearchTerm}</mark></b>`;
      hits.forEach((it, idx) => {
        hits[idx].markTitle = !hits[idx].title ? '' : hits[idx].title.replace(regex, replacement.toLowerCase());
      });
      setData(hits);
    }
  }, [status, debouncedSearchTerm]);

  return (
    <Wrapper className={active ? 'active' : ''}>
      <div
        ref={wrapperRef}
        className='search-input'
        onFocus={handleFocus}
      >
        <input
          value={value}
          onChange={handleChange}
          type='text'
          placeholder='Type and choose name of city...'
          id='search-input'
          autoComplete='off'
        />
        {debouncedSearchTerm  && active && (
          <div className='search-dropdown' id='hits-container'>
            {status === 'success' && data.length === 0 && <div>No location match with your input</div>}
            {status === 'pending' && <Spinner />}
            {data.length > 0 && status === 'success' && (
            <div className='ais-hits'>
              {data.map(item => (
                <div key={`location_${item.woeid}`} onClick={() => handleSelect(item)} className='ais-hits--item'>
                  <div className='hit'>
                    <div className='hit-content'>
                      <p className='hit-description' dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(item.markTitle) }} />
                    </div>
                  </div>
                </div>
                ))}
            </div>
            )}
          </div>
        )
        }
        <span onClick={handleReset}>
          <i className='close-search'>x</i>
        </span>
      </div>
    </Wrapper>
  );
};

SearchInput.propTypes = {
  onSelect: PropTypes.func,
};

SearchInput.defaultProps = {
  onSelect: () => {},
};

export default SearchInput;
