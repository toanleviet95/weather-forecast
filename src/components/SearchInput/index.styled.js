
import styled from 'styled-components';

export const Wrapper = styled.div`
flex-grow: 1;
-ms-flex-positive: 1;
position: relative;
height: 47px;
margin-top: 50px;
&.active {
  box-shadow: 0 4px 4px rgba(79, 79, 79, 0.21);
  .search-dropdown {
    display: block;
  }
  .search-input {
    -webkit-animation: expand-search-box-animation 0.5s forwards;
    animation: expand-search-box-animation 0.5s forwards;
    input {
      border-width: 2px;
    }
    .close-search {
      display: inline-block;
    }
  }
}
.search-input {
  transition: left 0.2s ease-in-out;
  transition: width 0s ease-in-out;
  .search-icon {
    position: absolute;
    left: 15px;
    top: 13px;
    z-index: 999;
    color: #bfc1c8;
  }
  input {
    font: 16px/1.875 "Avenir Next W01", "Avenir Next", "Helvetica Neue", Helvetica, sans-serif;
    height: 50px;
    border: 1px solid #bfc1c8;
    border-radius: 4px;
    width: 100%;
    padding-left: 10px;
    &:focus {
      outline: none;
    }
  }
  i.close-search {
    color: ##bfc1c8;
    display: none;
    position: absolute;
    right: 15px;
    top: 13px;
    cursor: pointer;
  }
  .search-dropdown {
    position: absolute;
    z-index: 999999999;
    box-sizing: border-box;
    color: #B3B3B3;
    font: 14px/1.875 "Avenir Next W01", "Avenir Next", "Helvetica Neue", Helvetica, sans-serif;
    opacity: 1.00;
    padding: 20px;
    width: 100%;
    -webkit-animation: expand-search-dropdown-animation 0.5s forwards;
    animation: expand-search-dropdown-animation 0.5s forwards;
    overflow-y: scroll;
    max-height: 400px;
    border-radius: 0 0 4px 4px;
    background-color: #FCFCFC;
    border: 1px solid #E0E0E0;
    box-shadow: 1px 3px 4px rgba(0, 0, 0, 0.09);
    .small {
      -webkit-flex-basis: 35%;
      -ms-flex-preferred-size: 35%;
      flex-basis: 35%;
    }
  }
}
.hit {
  border-bottom: 1px solid #E6E6E6;

  .hit-title {
    color: #bfc1c8;
    font-family: 'bt_mono', monospace;
    border-bottom: 1px solid #bfc1c8;
    font-weight: 500;
    margin-bottom: 0;
    margin-top: 0;
    display: inline-block;
    font-size: 14px;
  }
  .hit-description {
      text-decoration: none;
      color: black;
      font-size: 14px;
      display: block;
    margin-top: 3px;
  }
  .hit-anchor {
      font-size: 13px;
      color: #666;
  }
  &:last-child {
    margin-bottom: 0;
  }
}
.ais-hits--item {
  cursor: pointer;
  &:last-child {
    .hit {
      border: 0;  
    }
  }
  &:hover {
    background-color: #00BFFF;
  }
}
`;

export default Wrapper;
