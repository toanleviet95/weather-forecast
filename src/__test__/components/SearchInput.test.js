import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import SearchInput from '@/components/SearchInput';

const wait = () => new Promise(resolve => setTimeout(resolve, 2000));

describe('SearchInput', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<SearchInput />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show correct value when type', () => {
    const search = shallow(<SearchInput />);

    search.find('input').simulate('change', { target: { value: 'london' } });

    expect(search.find('input').props().value).toEqual('london');
  });

  it('should show dropdown when type correct location', () => {
    const search = mount(<SearchInput />);

    search.find('input').simulate('change', { target: { value: 'london' } });

    expect(search.find('input').props().value).toEqual('london');

    wait().then(() => {
      expect(search.find('.search-dropdown')).toHaveLength(1);
    });
  });
});
