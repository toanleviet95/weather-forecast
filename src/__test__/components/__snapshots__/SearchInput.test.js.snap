// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput renders correctly 1`] = `
<div
  className="sc-gtssRu ecHCLW"
>
  <div
    className="search-input"
    onFocus={[Function]}
  >
    <input
      autoComplete="off"
      id="search-input"
      onChange={[Function]}
      placeholder="Type and choose name of city..."
      type="text"
      value=""
    />
    
    <span
      onClick={[Function]}
    >
      <i
        className="close-search"
      >
        x
      </i>
    </span>
  </div>
</div>
`;
